### PR TITLE
[core] Fix alignment issues in cmsg header managment (#2830)

### PR DESCRIPTION
**Pull request by yomnes0**
We used to directly cast the result of CMSG_DATA to in_pktinfo or in6_pktinfo. As stated in CMSG_FIRSTHDR manpage, alignement of the return value of this method is not guaranteed, and a memcpy should be done to manipulate it. 

For some reason, the use of the mutable m_acCmsgRecvBuffer and m_acCmsgSendBuffer also caused alignemnt issues when using it as msg_control field of msghdr. Declaring a new char each time instead of using the mutable one solved this issue.

### CODE REVIEW

File: srtcore/channel.h [hunk 1]
```diff
@@ -236,14 +227,16 @@ class CChannel
             // IPv4 headers or IPv6 headers.
             if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_PKTINFO)
             {
-                in_pktinfo *dest_ip_ptr = (in_pktinfo*)CMSG_DATA(cmsg);
-                return sockaddr_any(dest_ip_ptr->ipi_addr, 0);
+                in_pktinfo dest_ip_ptr;
+                memcpy(&dest_ip_ptr, CMSG_DATA(cmsg), sizeof(struct in_pktinfo));
```
Comment (ethouris) [inline srtcore/channel.h:231]:
```suggestion
                in_pktinfo dest_ip = *(in_pktinfo*)CMSG_DATA(cmsg);
```
And BTW, what is this change for? Is that because CMSG_DATA doesn't guarantee to return an even pointer?

File: srtcore/channel.h [hunk 2]
```diff
@@ -236,14 +227,16 @@ class CChannel
...
+                return sockaddr_any(dest_ip_ptr.ipi_addr, 0);
             }
 
             if (cmsg->cmsg_level == IPPROTO_IPV6 && cmsg->cmsg_type == IPV6_PKTINFO)
             {
-                in6_pktinfo* dest_ip_ptr = (in6_pktinfo*)CMSG_DATA(cmsg);
-                return sockaddr_any(dest_ip_ptr->ipi6_addr, 0);
+                in6_pktinfo dest_ip_ptr;
+                memcpy(&dest_ip_ptr, CMSG_DATA(cmsg), sizeof(struct in6_pktinfo));
+                return sockaddr_any(dest_ip_ptr.ipi6_addr, 0);
```
Comment (ethouris) [inline srtcore/channel.h:239]: idem

File: srtcore/channel.h [hunk 3]
```diff
@@ -255,7 +248,7 @@ class CChannel
     // calling ::sendmsg function. It uses a static buffer to supply data
     // for the call, and it's stated that only one thread is trying to
     // use a CChannel object in sending mode.
-    bool setSourceAddress(msghdr& mh, const sockaddr_any& adr) const
+    bool setSourceAddress(msghdr& mh, void *buf, const sockaddr_any& adr) const
```
Comment (ethouris) [inline srtcore/channel.h:251]: If the variable passed here is a local variable created in the calling function, isn't it simpler to simply create this buffer inside this very function?

BTW. `void*` is "transparent" in C++, so use the original type and do not cast.

File: srtcore/channel.h [hunk 4]
```diff
@@ -264,30 +257,33 @@ class CChannel
 
         if (adr.family() == AF_INET)
         {
-            mh.msg_control = m_acCmsgSendBuffer;
-            mh.msg_controllen = CMSG_SPACE(sizeof(in_pktinfo));
-            cmsghdr* cmsg_send = CMSG_FIRSTHDR(&mh);
+            mh.msg_control = (void *) buf;
+            mh.msg_controllen = sizeof(struct in_pktinfo) + sizeof(cmsghdr);
```
Comment (ethouris) [inline srtcore/channel.h:261]: If you move the creation of the local variable to here, you can simply use `Size(buf)`.

File: srtcore/channel.cpp [hunk 5]
```diff
@@ -862,14 +863,15 @@ srt::EReadStatus srt::CChannel::recvfrom(sockaddr_any& w_addr, CPacket& w_packet
...
+        char mh_crtl_buf[sizeof(CMSGNodeIPv4) + sizeof(CMSGNodeIPv6)];
         if (m_bBindMasked)
         {
             // Extract the destination IP address from the ancillary
             // data. This might be interesting for the connection to
             // know to which address the packet should be sent back during
             // the handshake and then addressed when sending during connection.
-            mh.msg_control = (m_acCmsgRecvBuffer);
-            mh.msg_controllen = sizeof m_acCmsgRecvBuffer;
+            mh.msg_control = mh_crtl_buf;
```
Comment (ethouris) [inline srtcore/channel.cpp:873]: Maintain parentheses. This is a language-transparent marker pointing out that this variable will be used for writing.

File: srtcore/channel.cpp [hunk 6]
```diff
@@ -862,14 +863,15 @@ srt::EReadStatus srt::CChannel::recvfrom(sockaddr_any& w_addr, CPacket& w_packet
...
+            mh.msg_controllen = (sizeof(mh_crtl_buf));
```
Comment (ethouris) [inline srtcore/channel.cpp:874]: Parentheses are not required here at all. Just `sizeof mh_ctrl_buf`.

File: srtcore/channel.h [hunk 7]
```diff
@@ -236,14 +227,16 @@ class CChannel
             // IPv4 headers or IPv6 headers.
             if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_PKTINFO)
             {
-                in_pktinfo *dest_ip_ptr = (in_pktinfo*)CMSG_DATA(cmsg);
-                return sockaddr_any(dest_ip_ptr->ipi_addr, 0);
+                in_pktinfo dest_ip_ptr;
```
Comment (ethouris) [inline srtcore/channel.h:230]: Remove `_ptr`, this variable isn't a pointer.

File: srtcore/channel.h [hunk 8]
```diff
@@ -264,30 +257,33 @@ class CChannel
...
         if (adr.family() == AF_INET)
         {
-            mh.msg_control = m_acCmsgSendBuffer;
-            mh.msg_controllen = CMSG_SPACE(sizeof(in_pktinfo));
-            cmsghdr* cmsg_send = CMSG_FIRSTHDR(&mh);
+            mh.msg_control = (void *) buf;
+            mh.msg_controllen = sizeof(struct in_pktinfo) + sizeof(cmsghdr);
 
+            cmsghdr* cmsg_send = CMSG_FIRSTHDR(&mh);
             // after initializing msghdr & control data to CMSG_SPACE(sizeof(struct in_pktinfo))
```
Comment (ethouris) [inline srtcore/channel.h:264]: I think you can just as well remove this comment, especially that it's actually not true.

File: srtcore/channel.h [hunk 9]
```diff
@@ -264,30 +257,33 @@ class CChannel
...
             return true;
         }
 
         if (adr.family() == AF_INET6)
         {
-            mh.msg_control = m_acCmsgSendBuffer;
-            mh.msg_controllen = CMSG_SPACE(sizeof(in6_pktinfo));
-            cmsghdr* cmsg_send = CMSG_FIRSTHDR(&mh);
+            mh.msg_control = buf;
+            mh.msg_controllen = sizeof(struct in6_pktinfo) + sizeof(cmsghdr);
```
Comment (ethouris) [inline srtcore/channel.h:280]: Why this change?

**Review by ethouris — COMMENTED** [review]
Reviewed; some things below require explanation.

File: srtcore/channel.h (hunk 3 above)
Comment (yomnes0) [reply srtcore/channel.h:251]: That would not work. Inside this function, mh.msg_control is set to buf. If this was declared in this scope, it would be destroyed when exiting the function, and the information set to the cmsghdr would be lost. 
I thought about moving `mh.msg_control = (void *) buf;` to outside of setSourceAddress, but that didn't feel right

Comment (ethouris) [reply srtcore/channel.h:251]: Ah, right, missed that. So yes, it should then stay as is, but the type of the `buf` parameter should be `char*`, as the original one.

File: srtcore/channel.h (hunk 4 above)
Comment (ethouris) [reply srtcore/channel.h:261]: Provided you don't, though, why it doesn't use `CMSG_SPACE`?

Comment (yomnes0) [reply srtcore/channel.h:261]: I followed some examples I found in [there](https://man.archlinux.org/man/core/man-pages/CMSG_FIRSTHDR.3.en) (CMSG_FIRSTHDR manpage). They are setting controllen to `msg.msg_controllen = sizeof(u.buf);`, where u.buf is the equivalent of the control buffer I'm using

Comment (ethouris) [reply srtcore/channel.h:261]: But this buffer is declared as:
```
    char buf[CMSG_SPACE(sizeof(myfds))];
```
so sizeof on the whole buffer could be treated as an alias in this case.

Comment (yomnes0) [reply srtcore/channel.h:261]: At one point when I was trying to fix this issue I used 2 different buffers, one for IPv4 and another for IPv6. Both of them used CMSG_SPACE in their declaration. That's why I didn't use it there, but you are definitively right I'll fix that

Comment (ethouris) [reply srtcore/channel.h:261]: I did it this way initially, but then it turned out to be a bad idea. I found cases when at least theoretically the same CMSG structure may contain entries for both IPv4 and IPv6 pktinfo and then both are found.

See also the comment in `channel.h` line 194.

Comment (yomnes0) [reply srtcore/channel.h:261]: Not using CMSG_SPACE caused also another issue that you asked me about in another comment. I added `sizeof(cmsghdr)` to avoid CMSG_DATA returning NULL. Thanks for your input, I'll push something that address all of it soon

**Review by ethouris — APPROVED** [review]

**Review by maxsharabayko — APPROVED** [review]
